docs/proposals/20210222-kubelet-authentication.md

```diff
@@ -0,0 +1,322 @@
+| node-attestation-controller | Part of Cluster API repository, under bootstrap/node/attestation, and imported by Cluster API infra providers for implementation. A generic challenge-response implementation will be included for providers / bare metal without an attestation mechanism. | A controller to verify and sign the CSR. This would be typically an importable controller where the infrastructure provider implements the interface with specific code for CSR <expand this> approval and start the controller as part of its main.go or through an independent binary. |
+
+#### Kubelet authentication plugin
+
+![sequence diagram](images/kubelet-authentication/bootstrap-flow.png)
+The following sequence diagram shows at a high level how a nodeConfig enables bootstrapping a Kubernetes node. This sequence diagram is independent of API definition and alternatives and is relevant in all cases.
+
+We propose a kubelet authentication plugin to be present on the instances, (the nodeadm CLI will be baked into the machine images through image-builder), which will be responsible for node registration, as well as certificate rotation. The agent will be made up of two parts:
+- A common library vendored from Cluster API which includes the following functionality:
+    - Certificate filesystem locking
+    - Checking existing certificate validity
+    - Certificate signing request generation for both kubelet client and serving certificates
+    - Submission of CSRs to the API server and waiting for approval
+- A provider specific implementation for node attestation
+    - A provider will only need to implement the generation of the attestation to be included in the CSR and the retrieval of the provider ID to be stored in an X.509 extension attribute.
+
+The behaviour of the authentication plugin will be as follows:
+
+
+#### Node Attestation
+
+As for the node-attestation-controller, the following interface needs to be implemented by the infrastructure providers:
+```go
+type ClusterAPISigner interface {
+     VerifyClientAttestationData (csr *certificatesv1beta1.CertificateSigningRequest) err
+     VerifyServingAttestationData (csr *certificatesv1beta1.CertificateSigningRequest) err
+}
+```
+
+This enables infrastructure providers to perform infrastructure-specific validation of node attestations (TPM, appended tags by the provider, etc.)
+
+Cluster API is responsible for partially verifying node identity with the following conditions:
+
+- A corresponding machine object exist for the `.spec.Username` (`system:nodes:<nodename>`)
```

Comment:
With this conditions would it be a concern that as a malicious entity I generate enough workload as to trigger autoscaling which results in "infinite" number of new machines created so I can expect one nodeName to match my malicious handcrafted CSRs and therefore impersonating a node?

Reply:
Infra provider will still need to verify host attestation and should ultimately reject it, so we'd defer it to a infra provider concern, but it is a valid consideration.

Reply:
Added statement